### PR TITLE
Add recommendations for asynchronous processing

### PR DESCRIPTION
**Pull request by benoittgt**
Started a discussion on discourse about it  : http://discuss.jsonapi.org/t/long-running-jobs-or-asynchronous-processing/26

Maybe the status response have too much infos for this example.
Maybe "actions" should not be there.

Any feedbacks more than welcome.

### CODE REVIEW

**Comment by benoittgt** [comment]
Do I need to add some change ?

**Comment by ryansch** [comment]
At first glance, this is really good stuff.

File: recommendations/index.md [hunk 1]
```diff
@@ -149,3 +149,53 @@ Although JSON API does not specify the format of date and time fields, it is
...
+      },
+      {
+        "code": "invalid_format",
+        "field": "/photos/file01"
+      }
+      ]
+    },
+    "actions": {
+        "cancel": "/photos/queue/5234/cancel",
+        "delete": "/photos/queue/5234/delete"
```
Comment (ryansch) [inline recommendations/index.md]: What would be the difference between `cancel` and `delete`?

**Comment by ethanresnick** [comment]
@benoittgt Coming back to this... I'd like to see this merged, but it needs a few updates. I'll comment inline. Let me know if you're up for bringing this up to date.

File: recommendations/index.md [hunk 2]
```diff
@@ -149,3 +149,53 @@ Although JSON API does not specify the format of date and time fields, it is
...
+        "code": "too_long",
+        "field": "/photos/title"
+      },
+      {
+        "code": "invalid_format",
+        "field": "/photos/file01"
+      }
+      ]
+    },
+    "actions": {
```
Comment (ethanresnick) [inline recommendations/index.md]: First, this `"actions"` container should (temporarily) be removed. We're working on a standard way to support actions (see, e.g., #745) but, until then, I don't want to recommend something that will likely contradict our final solution.

File: recommendations/index.md [hunk 3]
```diff
@@ -149,3 +149,53 @@ Although JSON API does not specify the format of date and time fields, it is
...
+
+## Asynchronous Processing <a href="#asynchronous-processing" id="asynchronous-processing" class="headerlink"></a>
+
+Consider a situation when you need to create a resource and the operation takes long time to complete.
+
+The request **SHOULD** return a status `202 Accepted` with a link in Header/Location.
+
+```text
+HTTP/1.1 202 Accepted
+Location: https://example.com/photos/queue/5234
```
Comment (ethanresnick) [inline recommendations/index.md]: Fwiw, HTTP [does not define](https://tools.ietf.org/html/rfc7231#section-7.1.2) the meaning of the `Location` header for a `202` response, so it's not clear that this is appropriate here. (It only defines it for 201 and 3xx.) A more appropriate header might be `Link` with some relation. I'll need to think about this. 

//cc @steveklabnik

File: recommendations/index.md [hunk 4]
```diff
@@ -149,3 +149,53 @@ Although JSON API does not specify the format of date and time fields, it is
...
+
+```json
+{
+  "links": {
+    "self": "/photos/queue/5234"
+  },
+    "data": {
+      "type": "queue job",
+      "id": "5234",
+      "status": "Process failed, see errors",
```
Comment (ethanresnick) [inline recommendations/index.md]: `"status"` and `"errors"` need to be in an `"attributes"` bucket to comply with JSON API.

File: recommendations/index.md (hunk 1 above)
Comment (benoittgt) [inline recommendations/index.md]: A job canceled can be restarted. But it's ok for me if a job is only deleted.

File: recommendations/index.md (hunk 2 above)
Comment (benoittgt) [inline recommendations/index.md]: I'm ok with that

File: recommendations/index.md (hunk 4 above)
Comment (benoittgt) [inline recommendations/index.md]: Yep. Thanks.

File: recommendations/index.md (hunk 3 above)
Comment (benoittgt) [inline recommendations/index.md]: Yes you right. Maybe the link in `links` in enough ?

File: recommendations/index.md [hunk 5]
```diff
@@ -149,3 +149,51 @@ Although JSON API does not specify the format of date and time fields, it is
...
+{
+   "links":{
+      "self":"/photos/queue/5234"
+   },
+   "data":{
+      "type":"queue job",
+      "id":"5234",
+      "attributes":{
+         "status":"Process failed, see errors",
+         "errors":[
```
Comment (ethanresnick) [inline recommendations/index.md]: These errors shouldn't be here, I don't think, because if the server already knows that the request has errored, it should send back the error (not a 200).

File: recommendations/index.md (hunk 3 above)
Comment (ethanresnick) [inline recommendations/index.md]: Yes, a link in `links` could be enough, but I don't think a top-level `self` link is the right link. The top-level `"self"` is defined as the URI the request was made to. So, the top-level `self` (if present at all—and, in fact, it probably doesn't make sense to include in response to a POST) would be the URI the user sent the job to, not the URI of the queued job itself.

A resource-level `self` link could work, though. So the request would be `POST /tasks` and the response would be:

``` text
HTTP/1.1 202 Accepted
Content-Type: application/vnd.api+json
```

``` js
{
   "data": {
      "type": "queued",
      "id": "1234",
      "attributes": { "status": "pending..." },
      "links": { "self": "/photos/queue/5234"} // <- the link would go here
   }
}
```

Does that work for you?

It seems to very nicely comply with HTTP, which says that:

> The representation sent with [a 202] response ought to describe the request's current status and point to (or embed) a status monitor that can provide the user with an estimate of when the request will be fulfilled.

Comment (ethanresnick) [inline recommendations/index.md]: Ok, I figured it out! The 202 response should contain a `Content-Location` header that points to the URI of the queue/status resource. From the [HTTP spec](https://tools.ietf.org/html/rfc7231#section-3.1.4.2):

> If Content-Location is included in a 2xx (Successful) response... and its field-value refers to a URI that differs from the effective request URI, then the origin server claims that the URI is an identifier for a different resource corresponding to the enclosed representation... [For a POST request, among others], such a Content-Location indicates that this payload is a representation reporting on the requested action's status and that the same report is available (for future access with GET) at the given URI."

Everything above I said about using a resource-level self link still applies, but `Content-Location` should be sent as well.

File: recommendations/index.md (hunk 5 above)
Comment (benoittgt) [inline recommendations/index.md]: Ok. Managing errors seems not related to this recommendation. Errors where discussed here https://github.com/json-api/json-api/pull/828. Seems enough for me.

File: recommendations/index.md (hunk 3 above)
Comment (benoittgt) [inline recommendations/index.md]: More than ok for me.

File: recommendations/index.md [hunk 6]
```diff
@@ -149,3 +149,41 @@ Although JSON API does not specify the format of date and time fields, it is
 recommended that servers align with ISO 8601. [This W3C
 NOTE](http://www.w3.org/TR/NOTE-datetime) provides an overview of the
 recommended formats.
+=======
+
+## Asynchronous Processing <a href="#asynchronous-processing" id="asynchronous-processing" class="headerlink"></a>
+
+Consider a situation when you need to create a resource and the operation takes long time to complete.
+
+The request **SHOULD** return a status `202 Accepted` with a link in Header/Location.
```
Comment (ethanresnick) [inline recommendations/index.md]: This should now read "... with a URI in the `Content-Location` header pointing to the resource describing the operation's  status".

File: recommendations/index.md [hunk 7]
```diff
@@ -149,3 +149,41 @@ Although JSON API does not specify the format of date and time fields, it is
...
+Content-Type: application/vnd.api+json
+Content-Location: https://example.com/photos/queue/5234
+```
+
+The location gave **SHOULD** return `200 Ok` with the status of the selected job. The body **MAY** contains `"actions"` to manage the job.
+
+```json
+{
+   "links":{
+      "self":"/photos/queue/5234"
```
Comment (ethanresnick) [inline recommendations/index.md]: As discussed, this top-level `self` link should go away

File: recommendations/index.md [hunk 8]
```diff
@@ -149,3 +149,41 @@ Although JSON API does not specify the format of date and time fields, it is
...
+
+The request **SHOULD** return a status `202 Accepted` with a link in Header/Location.
+
+```text
+HTTP/1.1 202 Accepted
+Content-Type: application/vnd.api+json
+Content-Location: https://example.com/photos/queue/5234
+```
+
+The location gave **SHOULD** return `200 Ok` with the status of the selected job. The body **MAY** contains `"actions"` to manage the job.
```
Comment (ethanresnick) [inline recommendations/index.md]: This text can now be removed too :)

File: recommendations/index.md [hunk 9]
```diff
@@ -149,3 +149,41 @@ Although JSON API does not specify the format of date and time fields, it is
...
+      "type":"queue job",
+      "id":"5234",
+      "attributes": { "status": "Pending request, waiting other process" },
+      "links": { "self": "/photos/queue/5234"}
+      }
+   }
+}
+```
+
+When job process is done the location gave **SHOULD** return `303 See other` with the location of the data in Header/Location.
```
Comment (ethanresnick) [inline recommendations/index.md]: s/gave/given earlier/

s/Header\/Location/its `Location` header.

**Comment by ethanresnick** [comment]
@benoittgt Awesome! I just noted some small grammatical corrections too. Can you please make those, squash this all down to one commit, and then I'll merge it?

**Comment by benoittgt** [comment]
Thanks Ethan for the time you took to review my PR. :sake:

**Comment by ethanresnick** [comment]
Happy to :) This looks great now. Merged!

**Comment by krainboltgreene** [comment]
Should this also mention that further requests should return `102`?
